src/OpenKNX/Time/TimeManager.cpp

```diff
@@ -0,0 +1,732 @@
+                result = std::string(timezoneString).substr(0, seperator - timezoneString + 1);
+
+            if (daylightSavingMode == DaylightSavingMode::AlwaysDayLightSavingTime)
+                result += "0,366"; // Always daylight saving tiem
+            else
+                result += "366,367"; // Always standard time
+            return result;
+        }
+
+        void TimeManager::sendTime()
+        {
+            _lastSendSecond = 255;
+        }
+
+        void TimeManager::loop()
+        {
+            if (_timeProvider != nullptr)
+                _timeProvider->loop();
+
+            if (ParamBASE_InternalTime && isTimeValid())
+            {
+                tm localTime = getLocalTime();
+                if (localTime.tm_sec != _lastSendSecond || localTime.tm_min != _lastSendMinute || localTime.tm_hour != _lastSendHour)
+                {
+                    bool forceSend = _lastSendSecond == 255;
+                    _lastSendSecond = localTime.tm_sec;
+                    _lastSendMinute = localTime.tm_min;
+                    _lastSendHour = localTime.tm_hour;
+
+                    // update time KO
+                    tm knxTime;
+                    knxTime.tm_hour = localTime.tm_hour;
+                    knxTime.tm_min = localTime.tm_min;
+                    knxTime.tm_sec = localTime.tm_sec;
+                    KoBASE_Time.valueNoSend(knxTime, DPT_TimeOfDay);
+
+                    // update date KO
+                    tm knxDate;
+                    knxDate.tm_year = localTime.tm_year + 1900;
+                    knxDate.tm_mon = localTime.tm_mon + 1;
+                    knxDate.tm_mday = localTime.tm_mday;
+                    knxDate.tm_wday = localTime.tm_wday;
+                    if (knxDate.tm_wday == 0)
+                        knxDate.tm_wday = 7;
+                    KoBASE_Date.valueNoSend(knxDate, DPT_Date);
+
+                    // update date/time KO
+                    knxDate.tm_hour = localTime.tm_hour;
+                    knxDate.tm_min = localTime.tm_min;
+                    knxDate.tm_sec = localTime.tm_sec;
+
+                    KoBASE_DateTime.valueNoSend(knxDate, DPT_DateTime);
+
+                    uint8_t* raw = KoBASE_DateTime.valueRef();
+                    if (localTime.tm_isdst)
+                        raw[6] |= DPT19_SUMMERTIME;
+                    else
+                        raw[6] &= ~DPT19_SUMMERTIME;
+
+                    raw[6] &= ~DPT19_WORKING_DAY; // not supported, always 0
+                    raw[6] |= DPT19_NO_WORKING_DAY;
+
+                    raw[6] &= ~DPT19_FAULT;
+                    raw[6] &= ~DPT19_NO_YEAR;
+                    raw[6] &= ~DPT19_NO_DATE;
+                    raw[6] &= ~DPT19_NO_DAY_OF_WEEK;
+                    raw[6] &= ~DPT19_NO_TIME;
+
+                    KoBASE_IsSummertime.valueNoSend(localTime.tm_isdst, DPT_Switch);
+
+                    if ((_lastSendMinute % 10 == 0 && _lastSendSecond == 0) || forceSend)
```

Comment:
Ob auf den Bus gesendet wird, sollte konfigurierbar sein.

Ein Senden an Minutengrenzen begünstigt Peaks in der Bus-Last, weil andere Aktionen auch oft am Minutenwechsel ausgeführt werden. Sendeintervall sollte unabhängig von der aktuell ausgegebenen Uhrzeit sein, und stattdessen nur vom zeitlichen Abstand zum letzten Sendens (basierend auf Uptime) abhängen. Bei zu langsam laufender Uhr und externem Setzen könnte es sonst im Extremfall passieren, dass der Sendezeitpunkt nie erreicht wird...

Reply:
Ein Zeitprovider ohne senden auf den Bus ergibt im Normalfall keinen Sinn, wenn man es in Ausnahmefällen nicht will, muss man die KO's nicht anschließen. Ich würde als Anwender sehr verwirrt sein, wenn ich NTP oder eine GPS Uhr verwende und dann die KO's noch konfigurieren müsste.

Ein Senden der Uhrzeit nicht an den Minutengrenzen würde eine viel aufwendigere Logik brauchen weil man dann die Zeitumstellung extra berücksichtigen müsste. Ich fürchte auch keine Peaks am Bus da es ja ohnedies nur einen Zeitgeber im KNX Netz geben sollte.

"Bei zu langsam laufender Uhr und externem Setzen könnte es sonst im Extremfall passieren, dass der Sendezeitpunkt nie erreicht wird..."
Wenn die Uhr so langsam läuft das innerhalb von 10 Minuten die Abweichung so groß wird, dann hast du ein Hardwareproblem.

Reply:
das muss weil jederzeit ein lesen eintreffen könnte

Reply:
> das muss weil jederzeit ein lesen eintreffen könnte
Geht mir hier nur um das Senden im dort beginnenden if-Block, `valueNoSend()` braucht es natürlich immer. Ist hier von der Darstellung allerdings auch etwas unglücklich, dass nur die Zeilen darüber sichtbar sind und nicht die darunter...

Reply:
> Ein Senden der Uhrzeit nicht an den Minutengrenzen würde eine viel aufwendigere Logik brauchen weil man dann die Zeitumstellung extra berücksichtigen müsste.

Meinst Du weil Du bei DST-Änderung immer sofort senden willst? (Was auf jeden Fall passieren sollte)

Eine relativ einfache Logik könnte so aussehen:
```
if (dstNew != dstOld)
  forceSend = true;

if (forceSend || delayCheckMillis(_lastTimeSend, ParamBASE_TimeSendInterval_ms))
{
  // ... send ...

  // reduce intervall after forced sending 
  _lastTimeSend = millis() - (forceSend ? 30*1000 : 0);
}
```



>Ich fürchte auch keine Peaks am Bus da es ja ohnedies nur einen Zeitgeber im KNX Netz geben sollte.

War auch nicht speziall auf Zeit-Telegramme bezogen.

Reply:
> War auch nicht speziall auf Zeit-Telegramme bezogen.
Sonst bin ich da auch dabei, aber gerade Zeittelegramme finde ich besser genau bei der Zeit. Dein Code z.B. kann ziemliche Rundungsprobleme verursachen weil z.B. eventuell erst kurz vor Ende oder aber auch am Anfang einer Sekunde gesendet wird, mit meinem Code wird immer am Anfang gesendet, damit ist die besten Genauigkeit gewährleistet.

Reply:
> Dein Code z.B. kann ziemliche Rundungsprobleme verursachen weil z.B. eventuell erst kurz vor Ende oder aber auch am Anfang einer Sekunde gesendet wird, mit meinem Code wird immer am Anfang gesendet, damit ist die besten Genauigkeit gewährleistet.

Die Prüfung der Sende-Bedingung erfolgt in beiden Fällen innerhalb des KO-Update-Blocks, der bei erkannter Sekunden-Änderung
```
tm localTime = getLocalTime();
if (localTime.tm_sec != _lastSendSecond || localTime.tm_min != _lastSendMinute || localTime.tm_hour != _lastSendHour)
{
```
ausgeführt wird. In einem Zeitraum innerhalb "Gesamtdauer aller loop()"-Aufrufe (Soll <5ms) nach Sekundenwechsel, sofern ich nichts übersehe. Danach werden die KOs aktualisiert (<1ms erwartet; ohne Messung), bevor geprüft wird ob ein Senden auf den Bus erfolgen soll. D.h.: Das Senden wird immer innerhalb von 10ms nach Sekundenänderung angestoßen. Das tatsächliche Senden wird erst - später - in in einem Aufruf von knx.loop() erfolgen. Auf TP wird es etwa 100ms dauern bis alle 4 KOs gesendet wurden. Sofern nicht vorher noch andere KOs senden.

Die aktuelle Implementierend würde immer nur in Sekunde 0 der Minuten 0,10,20,30,40,50 senden. Müsste nun noch noch mal in die Implementierung von LOG schauen, ob die Zeitschaltuhren auch in Sekunde 0 senden (oder 1 wie irgendwo schon mal gesehen hatte). Dortiges Senden in Sekunde 0 würde konkurrieren mit Senden der Zeit.

Reply:
Entscheidung: Wir senden explizit an Minutengrenzen, da es Geräte gibt die beim Empfang von Zeitangaben die Sekunden auf 0 zurücksetzen.